esp/services/ws_workunits/ws_wudetails.hpp

```diff
@@ -0,0 +1,126 @@
+/*##############################################################################
+
+    HPCC SYSTEMS software Copyright (C) 2017 HPCC Systems®.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#ifndef _ESPWIZ_ws_wudetails_HPP__
+#define _ESPWIZ_ws_wudetails_HPP__
+
+#include "jlib.hpp"
+
+class CWUDetails
+{
+    Owned<IConstWorkUnit> workunit;
+    StringBuffer wuid;
+    IConstWUScopeFilter & scopeFilter;
+    IConstWUAttributeToReturn & attribsToReturn;
+    IConstWUScopeOptions & scopeOptions;
+    IConstWUAttributeOptions & attribOptions;
+    IConstWUNestedFilter & nfilter;
+
+    class CAttributeFilter: public CInterface, implements IInterface
+    {
+        StatisticKind statisticKind;
+        unsigned __int64 exactValue;
+        unsigned __int64 minValue;
+        unsigned __int64 maxValue;
+        bool bExactValue;
+        bool bMinValue;
+        bool bMaxValue;
+    public:
+        IMPLEMENT_IINTERFACE;
+
+        CAttributeFilter(){};
+        CAttributeFilter(StatisticKind sk, unsigned __int64 exact, unsigned __int64 min,
+                        unsigned __int64 max, bool bExact, bool bMin, bool bMax)
+                        : statisticKind(sk), exactValue(exact), minValue(min), maxValue(max),
+                        bExactValue(bExact), bMinValue(bMin), bMaxValue(bMax) {};
+        StatisticKind getStatisticKind() { return statisticKind; }
+        void setStatisticKind(StatisticKind sk) { statisticKind = sk; };
+        unsigned __int64 getExactValue() { return exactValue; }
+        unsigned __int64 getMinValue()   { return minValue; }
+        unsigned __int64 getMaxValue()   { return maxValue; }
+        bool hasExactValue()             { return bExactValue; }
+        bool hasMinValue()               { return bMinValue; }
+        bool hasMaxValue()               { return bMaxValue; }
+    };
+
+    static int funcCompareAttribFilter(IInterface* const * l, IInterface* const * r)
+    {
+        CAttributeFilter * ll = (dynamic_cast<CAttributeFilter *>(*l));
```

Comment:
dynamic casts should be avoided unless there is no alternative.  They are inefficient.